src/server-logger.ts

```diff
@@ -0,0 +1,257 @@
+      this.sendLogsToServer(
+        /* set to immediate if last request removed from the buffer, and we have more than one item in the buffer */
+        removeFromBuffer && this.logBuffer.length > 1
+      );
+    }
+  }
+
+  private sendAllLogsInstantly () {
+    /* if `sendAll`, we need to send all requests in the buffer without awaiting */
+    return this.logBuffer.forEach((item, index) => {
+      /* if we already have a pending request, it means index `0` already has been sent */
+      if (this.pendingHttpRequest && index === 0) {
+        return;
+      }
+      this.postLogsToEndpoint(this.config.url, item.traces.reverse());
+    });
+  }
+
+  private truncateLog (logLevel: LogLevel, log: ILogMessage): ITrace | null {
+    let trace: ITrace;
+    const logCopy = cloneDeep(log);
+    const truncText = '[[TRUNCATED]]';
+
+    /* first truncate the details */
+    logCopy.details = truncText;
+    trace = this.convertToTrace(logLevel, logCopy);
+
+    if (calculateLogMessageSize(trace) <= MAX_LOG_SIZE) {
+      this.logger.warn('message too large to send to server. truncated log details', {
+        originalLog: (log as any),
+        truncatedLog: (logCopy as any)
+      }, true);
+      return trace;
+    }
+
+    /* second truncate the message */
+    logCopy.message = `${logCopy.message.substr(0, 150)}... ${truncText}`;
+    trace = this.convertToTrace(logLevel, logCopy);
+
+    if (calculateLogMessageSize(trace) <= MAX_LOG_SIZE) {
+      this.logger.warn('message too large to send to server. truncated log details & log message', {
+        originalLog: (log as any),
+        truncatedLog: (logCopy as any)
+      }, true);
+      return trace;
+    }
+
+    /* if the truncated trace is _still_ too large, return null because we aren't going to send this to the server */
+    return null;
+  }
+
+  private convertToLogMessage (message: string, details: IDetail | IDetail[]): ILogMessage {
+    return {
+      clientTime: new Date().toISOString(),
+      clientId: this.logger.clientId,
+      message,
+      details
+    };
+  }
+
+  private convertToTrace (level: LogLevel, log: ILogMessage): ITrace {
+    return {
+      topic: this.config.logTopic,
+      level: level.toUpperCase(),
+      message: stringify(log)
+    };
+  }
+
+  private postLogsToEndpoint (url: string, traces: ITrace[]): Promise<any> {
+    // console.log('%c sending logs to server', 'color: #32a852', traces); // TODO: delete me
+
+    return request.post(url)
+      .set('Authorization', `Bearer ${this.config.accessToken}`)
+      .type('application/json; charset=UTF-8')
+      .send({
+        accessToken: this.config.accessToken,
+        app: {
+          appId: this.config.logTopic,
+          appVersion: this.config.appVersion
+        },
+        traces
+      }).then(() => {
+        // console.log('%c successfully sent logs to server', 'color: #32a852', { tracesSent: traces, currentBuffer: [...this.logBuffer] }); // TODO: delete me
```

Comment:
might be cool if we have a `debugMode` for the logger that just enables all these hidden messages instead of commenting them out.

Reply:
I like that idea. I will add it in.